File: mid/207. Course Schedule.py
# Time complexity: O(V+E)O(V + E)O(V+E)
# Space complexity: O(V+E)O(V + E)O(V+E)

class Solution:
    def canFinish(self, n: int, prerequisites: List[List[int]]) -> bool:
        adj = [[] for _ in range(n)]
        prerequisites_count = [0] * n
        ans = []

        for pair in prerequisites:
            course = pair[0]
            prerequisite = pair[1]
            adj[prerequisite].append(course)
            prerequisites_count[course] += 1
            #print('adj=', adj)
            #print('prerequisites_count=', prerequisites_count)

        queue = deque()
        for i in range(n):
            if prerequisites_count[i] == 0:
                queue.append(i)

        while queue:
            current = queue.popleft()
            ans.append(current)

            for next_course in adj[current]:
                prerequisites_count[next_course] -= 1
                if prerequisites_count[next_course] == 0:
                    queue.append(next_course)

        return len(ans) == n

File: mid/207. Course Schedule.cpp
// Time complexity: O(V+E)O(V + E)O(V+E)
// Space complexity: O(V+E)O(V + E)O(V+E)

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        int n = numCourses;
        unordered_set<int> can_take;
        vector<int> prerequisites_count(n, 0);
        vector<vector<int>> reverse_adj(n);
        
        // Create the adjacency lists and reversed adjacency lists
        for (auto pre : prerequisites) {
            reverse_adj[pre[0]].push_back(pre[1]);
            prerequisites_count[pre[1]] += 1;
        }
        
        queue<int> q;
        
        // Enqueue courses with no prerequisites
        for (int i = 0; i < n; i++) {
            if (prerequisites_count[i] == 0) {
                q.push(i);
            }
        }
        
        while (!q.empty()) {
            int course = q.front();
            q.pop();
            can_take.insert(course);
            
            for (auto postreq : reverse_adj[course]) {
                prerequisites_count[postreq] -= 1;
                if (prerequisites_count[postreq] == 0) {
                    q.push(postreq);
                }
            }
        }
        
        if (can_take.size() == n) {
            return true;
        } else {
            return false;
        }
    }
};